src/services/dataStreamer.ts

```diff
@@ -128,54 +209,15 @@ export const streamData = async (
   opts: DataStreamOptions = {},
   req: Request
 ) => {
-  const timestamp = Date.now();
   const { reqId } = opts;
+  const outputStream = chain([stringer(), res]);
+  // Mutable mapping to track what assets are used for different pages
+  const assetData = {};
 
-  res.once('error', (err) => {
-    logger.error(createMessage(`Error with response pipeline: ${err}`, reqId));
-    // Destroy streams in hopes of preventing memory leaks
-    res.destroy();
-  });
+  streamTimestamp(outputStream, reqId);
+  await streamMetadata(outputStream, metadataCursor, reqId);
+  await streamPages(outputStream, pagesCursor, assetData, opts);
+  await streamAssets(outputStream, assetData, req, reqId);
 
-  // Used as a chain of streams for output
-  const pipeline = chain([stringer(), res]);
-
-  // Return timestamp to inform Gatsby Cloud when data was last queried
-  const timestampChunk: StreamData = { type: 'timestamp', data: timestamp };
-  logger.info(createMessage(`Returning timestamp: ${timestamp}`, reqId));
-  pipeline.write(timestampChunk);
-
-  let metadataCount = 0;
-  const metadataStream = metadataCursor.stream({
-    transform(doc) {
-      const newDoc: StreamData = {
-        type: 'metadata',
-        data: doc,
-      };
-      metadataCount++;
-      return newDoc;
-    },
-  });
-
-  // We use pipe() instead of promisified pipeline() here due to weird behavior with
-  // having multiple streams in the same pipeline and setting `end` to false.
-  metadataStream.pipe(pipeline, { end: false });
-  // Chaining sequential streams like this is not ideal
-  metadataStream.once('end', async () => {
-    logger.info(createMessage(`Found ${metadataCount} metadata documents`, reqId));
-    try {
-      await streamPages(pipeline, pagesCursor, req, opts);
-    } catch (err) {
-      // Don't throw error since it'll just be a fatal error. Report it
```

Comment:
can we transport this why we're not throwing errors to handleStream fn

Reply:
I'm sorry, what do you mean by "can we transport"? Fwiw, I omitted this try catch block since I made the assumption that in the event of an error, the listener in [handleStream](https://github.com/mongodb/snooty-data-api/pull/42/files/a4ff02198a6eb051d0c417cb3618b66521d82377#diff-1b4341f0164335f21d23e9634e25b2af1e17f17bdf09d754dfd17b94226ce92dR49) can handle it. There's also a try catch block at the router level, which should hopefully handle logging any error

Reply:
oh, i meant the comment of why we're not throwing errors within the error listeners. i also saw that we aren't throwing from any of the `stream` functions ie. returned promise does not reject at all [here](https://github.com/mongodb/snooty-data-api/blob/a4ff02198a6eb051d0c417cb3618b66521d82377/src/services/dataStreamer.ts#L40) even though we are attempting to catch ([example](https://github.com/mongodb/snooty-data-api/blob/a4ff02198a6eb051d0c417cb3618b66521d82377/src/routes/projects.ts#L44))

Reply:
Oh, gotcha. I could actually try bringing up that error listener in `handleStream` and potentially use it as a way to propagate that error. Maybe something like:

```js
// Wait for stream to end
await new Promise<void>((resolve, reject) => {
  sourceStream.once('end', async () => {
    resolve();
  });

  sourceStream.once('error', async (err) => {
    logger.error(createMessage(`There was an error streaming ${dataType}: ${err}`, reqId));
    sourceStream.destroy();
    reject();
  });
});
```

Would that be more reasonable?

Reply:
that makes more sense to me !

Reply:
Done. Thanks for bringing this up!